### ISSUE TITLE
[BUG] fix noam scheduler definition

### ISSUE THREAD

**Issue by davidegraff** [chemprop/chemprop#969]
per #968, the actual definition of the Noam scheduler does not match up with the v1 implementation. It looks like it isn't necessary to know the total number of steps because it's originally defined as a square root decay rather than exponential decay:

```math
\nu = \sqrt{\frac{1}{n_\text{params}}} \cdot \min(\sqrt{t}, t / t_\text{warmup}^{−1.5})
```

So this seems to be an implementation error with the v1 scheduler and something we should fix.

**Comment by davidegraff** [comment]
FWIW, the $min$ operation effectively defines a piecewise function:

```math
\nu = \frac{1}{\sqrt{n_\text{params}}}
\begin{cases}
  \frac{t}{t_\text{warmup}^{-1.5}} &\text{for } 0 \leq t \leq t_\text{warmup}^{-1} \\
  \sqrt{t} &\text{otherwise}
\end{cases}\text{ ,}
```

where $t$ is the number of steps. So actually don't need a custom `NoamLRScheduler` class and can just define it as a [`ChainedScheduler`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ChainedScheduler.html#torch.optim.lr_scheduler.ChainedScheduler) consisting of a [`LinearLR`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LinearLR.html#torch.optim.lr_scheduler.LinearLR) and a [`LambdaLR`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LambdaLR.html#torch.optim.lr_scheduler.LambdaLR) together.

**Comment by davidegraff** [comment]
$d_\text{model}$ is the transformer hidden dimension (512) AIAYN, so for us it would the latent/message size (300). Also, we could define it your way or use a [`SequentialLR`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.SequentialLR.html#torch.optim.lr_scheduler.SequentialLR) which formally defines the piecewise function. I'm indifferent to the two, but any PR should provide a plot comparing the learning rates calculated via the old and new schedulers.

**Comment by davidegraff** [comment]
As far as the old implementation goes, yes, we should keep. I would honestly rename it to `IncorrectNoamLRScheduler` because that's what it is.

**Comment by shihchengli** [comment]
We discussed this in today's developer meeting. We think we could pass the scheduler into the `MPNN` object instead of creating the scheduler in the `configure_optimizers()` method of the `MPNN` object. Then, we will remove `warmup_epochs`, `init_lr`, `max_lr`, and `final_lr` as attributes in the `MPNN` in the next minor version, and it will be backward incompatible. We can also add a register in the schedulers.py to allow different scheduler options (for now, we can keep both the incorrect and correct Noam optimizers and add a flag in CLI to let users choose). We can also extend the options of schedulers in the future (e.g., cosine decay scheduler). Additionally, we decided to switch from `Adam` to `AdamW` as the optimizer, as it is a bug-fix version of Adam.

**Comment by davidegraff** [comment]
1) you can't simply pass in an a learning rate scheduler, as the pytorch method signature of schedulers is `Scheduler(opt: Otpimizer)`, which is to say you _can't_ (by definition) create a scheduler _before_ your optimizer. Your object model, proposed, isn't possible
2) AdamW is not a bug-fixed version of Adam, but just be sure to make very clear to users about this default change.

**Comment by shihchengli** [comment]
Thanks for pointing this out. How about passing a string to the `MPNN` and using a `get_lr_scheduler` function to obtain the learning rate scheduler?

**Comment by davidegraff** [comment]
the general problem you have here is that you want (a) a variable learning schedule that you can (b) parametrize as you wish. This is hard to achieve because _both_ of these details need to be passed into the model. I see two options:
1. pass two arguments to the model: `T_sched: type[LRScheduler]` and lr_sched_kwargs: Mapping. These are then used in configure_optimizers:

```python
def configure_optimizers(self):
    opt = …
    sched = self.T_sched(opt, **self.lr_sched_kwargs)
    …
```

2. alternatively, just pass in `lr_sched_factory: Callable[[Optimizer], LRScheduler]`. This is similar to (1) in practice:

```python
def configure_optimizers(self):
    opt = …
    sched = self.lr_sched_factory(opt)
    …
```

The practical difference between the two is that (2) is essentially (1) with a call to `partial` before passing in the argument if a user wants to parametrize their scheduler. I prefer (2) because it keeps things concise inside the model source by limiting the number of attributes/arguments dedicated to the scheduler.

**Comment by shihchengli** [comment]
Looks good! I will open a PR for this. In the AIAYN, when the current step is the number of warmup steps, the learning rate is defined as $d_{model}^{-0.5} \times {warmup\\_steps}^{-0.5}$, which is $512^{-0.5} \times 4000^{-0.5} = 7 \times 10^{-4}$. Instead of defining the model learning rate by the model size and the warmup steps, we can just use `init_lr` to replace this. That is to say, I think we can just provide an `init_lr` to the optimizer and, based on the Noam scheduler, we do a linear increase from `0` to `init_lr` in the warmup steps and then do an inverse square root decrease.

**Comment by davidegraff** [comment]
Honestly, the more I think about this, the less I think we actually need to implement the true Noam scheduler and more just refactor the one we currently have. IMO, it's a nice property of scheduler that it starts at `init_lr` and linearly warms to `max_lr` after `warmup_steps` and then exponentially cools to `final_lr` after `cooldown_steps`. The Noam scheduler doesn't have that property.

What I might suggest now is:
1.) We move over to defining this scheduler in relative terms, i.e., `warmup_factor` and `cooldown_factor`, rather than absolute terms as the former seems more intuitive to me.
2.) Define our scheduler using torch-native classes rather than a custom one.

Also, I still don't know what we should do in the case of infinite training. Perhaps just assume `cooldown_steps = 100*warmup_steps` and take the max of the calculated learning rate and `final_lr`? This way it ensures that our cooling schedule never cools beyond `final_lr`.

**Comment by shihchengli** [comment]
I am not sure how much benefit we can get from changing the scheduler, so I am fine with keeping the current approach. If we want to change the scheduler, I think other options might be better than the Noam optimizer, such as linear warmup with cosine decay or [ReduceLROnPlateau](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html). The benefit of using another scheduler is that we wouldn't have so many hyperparameters to optimize for the learning rate (we currently have `init_lr`, `max_lr`, `final_lr`, and `warmup_epochs`). I agree with your points on refactoring the code. I am not sure what would be the proper way to deal with the infinite. I think using the method you suggested is fine. On the other hand, do you still think we could provide some flexibility for people to define their user-defined schedulers (i.e., passing `lr_sched_factory` into MPNN)?

**Comment by davidegraff** [comment]
IMO it's unlikely that a simple GNN could materially benefit from a different LR scheduler. You might get a small bump in performance, but it's likely something you could also achieve by modifying other hyperparameters. I think it's a hopeful assumption that there are "better" schedulers out there, and adding them in would just serve to _increase_ our hyperparameter space. That said, I don't really care if you want to go about implementing this change, and it would be a good change for the users on the library side.
